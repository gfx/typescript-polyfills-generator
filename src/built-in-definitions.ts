const arrayNatureIterators: string[] = ['es6.array.iterator'];
const commonIterators: string[] = ['es6.string.iterator'].concat(arrayNatureIterators);

export interface Definitions {
    [key: string]: string[];
}

export interface GroupDefinitions {
    [key: string]: Definitions;
}

export const builtIns: Definitions = {
    fetch: ['es6.fetch'],
    DataView: ['es6.typed.data-view'],
    Float32Array: ['es6.typed.float32-array'],
    Float64Array: ['es6.typed.float64-array'],
    Int8Array: ['es6.typed.int8-array'],
    Int16Array: ['es6.typed.int16-array'],
    Int32Array: ['es6.typed.int32-array'],
    Map: ['es6.map'].concat(commonIterators),
    Number: ['es6.number.constructor'],
    Promise: ['es6.promise'],
    RegExp: ['es6.regexp.constructor'],
    Set: ['es6.set'].concat(commonIterators),
    Symbol: ['es6.symbol', 'es7.symbol.async-iterator'],
    Uint8Array: ['es6.typed.uint8-array'],
    Uint8ClampedArray: ['es6.typed.uint8-clamped-array'],
    Uint16Array: ['es6.typed.uint16-array'],
    Uint32Array: ['es6.typed.uint32-array'],
    WeakMap: ['es6.weak-map'].concat(commonIterators),
    WeakSet: ['es6.weak-set'].concat(commonIterators)
};

export const instanceMethods: GroupDefinitions = {
    Object: {
        __defineGetter__: ['es7.object.define-getter'],
        __defineSetter__: ['es7.object.define-setter'],
        __lookupGetter__: ['es7.object.lookup-getter'],
        __lookupSetter__: ['es7.object.lookup-setter']
    },
    String: {
        anchor: ['es6.string.anchor'],
        big: ['es6.string.big'],
        blink: ['es6.string.blink'],
        bold: ['es6.string.bold'],
        codePointAt: ['es6.string.code-point-at'],
        endsWith: ['es6.string.ends-with'],
        fixed: ['es6.string.fixed'],
        fontcolor: ['es6.string.fontcolor'],
        fontsize: ['es6.string.fontsize'],
        includes: ['es6.string.includes'],
        italics: ['es6.string.italics'],
        link: ['es6.string.link'],
        padStart: ['es7.string.pad-start'],
        padEnd: ['es7.string.pad-end'],
        repeat: ['es6.string.repeat'],
        small: ['es6.string.small'],
        startsWith: ['es6.string.starts-with'],
        strike: ['es6.string.strike'],
        sub: ['es6.string.sub'],
        trim: ['es6.string.trim']
    },
    Array: {
        copyWithin: ['es6.array.copy-within'],
        entries: arrayNatureIterators,
        every: ['es6.array.is-array'],
        fill: ['es6.array.fill'],
        filter: ['es6.array.filter'],
        find: ['es6.array.find'],
        findIndex: ['es6.array.find-index'],
        forEach: ['es6.array.for-each'],
        includes: ['es7.array.includes'],
        indexOf: ['es6.array.index-of'],
        keys: arrayNatureIterators,
        lastIndexOf: ['es6.array.last-index-of'],
        map: ['es6.array.map'],
        reduce: ['es6.array.reduce'],
        reduceRight: ['es6.array.reduce-right'],
        slice: ['es6.array.slice'],
        some: ['es6.array.some'],
        sort: ['es6.array.sort'],
        values: arrayNatureIterators
    },
    Function: {
        bind: ['es6.function.bind'],
        name: ['es6.function.name']
    },
    Promise: {
        finally: ['es7.promise.finally']
    },
    RegExp: {
        flags: ['es6.regexp.flags'],
        match: ['es6.regexp.match'],
        replace: ['es6.regexp.replace'],
        search: ['es6.regexp.search'],
        split: ['es6.regexp.split'],
        toString: ['es6.regexp.to-string']
    },
    Date: {
        toISOString: ['es6.date.to-iso-string'],
        toJSON: ['es6.date.to-json'],
        toString: ['es6.date.to-string']
    }
};

export const staticMethods: GroupDefinitions = {
    Array: {
        from: ['es6.array.from'],
        isArray: ['es6.array.is-array'],
        of: ['es6.array.of']
    },
    Date: {
        now: ['es6.date.now']
    },
    Object: {
        assign: ['es6.object.assign'],
        create: ['es6.object.create'],
        defineProperty: ['es6.object.define-property'],
        defineProperties: ['es6.object.define-properties'],
        entries: ['es7.object.entries'],
        freeze: ['es6.object.freeze'],
        getOwnPropertyDescriptors: ['es7.object.get-own-property-descriptors'],
        is: ['es6.object.is'],
        isExtensible: ['es6.object.is-extensible'],
        isFrozen: ['es6.object.is-frozen'],
        isSealed: ['es6.object.is-sealed'],
        keys: ['es6.object.keys'],
        preventExtensions: ['es6.object.prevent-extensions'],
        seal: ['es6.object.seal'],
        setPrototypeOf: ['es6.object.set-prototype-of'],
        values: ['es7.object.values']
    },
    Math: {
        acosh: ['es6.math.acosh'],
        asinh: ['es6.math.asinh'],
        atanh: ['es6.math.atanh'],
        cbrt: ['es6.math.cbrt'],
        clz32: ['es6.math.clz32'],
        cosh: ['es6.math.cosh'],
        expm1: ['es6.math.expm1'],
        fround: ['es6.math.fround'],
        hypot: ['es6.math.hypot'],
        imul: ['es6.math.imul'],
        log1p: ['es6.math.log1p'],
        log10: ['es6.math.log10'],
        log2: ['es6.math.log2'],
        sign: ['es6.math.sign'],
        sinh: ['es6.math.sinh'],
        tanh: ['es6.math.tanh'],
        trunc: ['es6.math.trunc']
    },
    String: {
        fromCodePoint: ['es6.string.from-code-point'],
        raw: ['es6.string.raw']
    },
    Number: {
        EPSILON: ['es6.number.epsilon'],
        MIN_SAFE_INTEGER: ['es6.number.min-safe-integer'],
        MAX_SAFE_INTEGER: ['es6.number.max-safe-integer'],
        isFinite: ['es6.number.is-finite'],
        isInteger: ['es6.number.is-integer'],
        isSafeInteger: ['es6.number.is-safe-integer'],
        isNaN: ['es6.number.is-nan'],
        parseFloat: ['es6.number.parse-float'],
        parseInt: ['es6.number.parse-int']
    },
    Promise: {
        all: commonIterators,
        race: commonIterators
    },
    Reflect: {
        apply: ['es6.reflect.apply'],
        construct: ['es6.reflect.construct'],
        defineProperty: ['es6.reflect.define-property'],
        deleteProperty: ['es6.reflect.delete-property'],
        get: ['es6.reflect.get'],
        getOwnPropertyDescriptor: ['es6.reflect.get-own-property-descriptor'],
        getPrototypeOf: ['es6.reflect.get-prototype-of'],
        has: ['es6.reflect.has'],
        isExtensible: ['es6.reflect.is-extensible'],
        ownKeys: ['es6.reflect.own-keys'],
        preventExtensions: ['es6.reflect.prevent-extensions'],
        set: ['es6.reflect.set'],
        setPrototypeOf: ['es6.reflect.set-prototype-of']
    }
};
