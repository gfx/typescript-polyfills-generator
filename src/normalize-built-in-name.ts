const versionPattern: RegExp = /^es[0-9]+\./;

export default function normalizeBuiltInName (name: string): string {
    return name.replace(versionPattern, '');
}