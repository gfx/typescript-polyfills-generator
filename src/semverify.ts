import {valid} from 'semver';

export const firstVersion: string = '0.0.1';

// Convert version to a semver value.
// 2.5 -> 2.5.0; 1 -> 1.0.0;
export default function semverify(version: string): string {
    if (!version) {
        return firstVersion;
    }

    if (valid(version)) {
        return version;
    }

    const parts: string[] = version.split('.');

    while (parts.length < 3) {
        parts.push('0');
    }

    return parts.join('.');
}
