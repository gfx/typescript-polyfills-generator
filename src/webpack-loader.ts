import {getOptions, OptionObject} from 'loader-utils';
import * as validateOptions from 'schema-utils';
import {loader as webpackLoader} from 'webpack';
import {PolyfillsGenerator, PolyfillsGeneratorOptions} from './index';

export interface LoaderOptions extends PolyfillsGeneratorOptions, OptionObject {
    //
}

const loaderName: string = 'ts-polyfill-loader';
const loaderOptionsSchema = {
    type: 'object',
    properties: {
        targets: {
            type: 'string'
        }
    },
    additionalProperties: false
};
const generators: {[key: string]: PolyfillsGenerator} = Object.create(null);

export default function loader (this: webpackLoader.LoaderContext, source: string) {
    const options: LoaderOptions = getOptions(this) as LoaderOptions;

    validateOptions(loaderOptionsSchema, options, loaderName);
    const optionsKey: string = JSON.stringify(options);
    const callback = this.async() as webpackLoader.loaderCallback;

    if (!generators[optionsKey]) {
        generators[optionsKey] = new PolyfillsGenerator(options);
    }

    const generator: PolyfillsGenerator = generators[optionsKey];

    generator.process(source).then(() => {

    })

    callback(null, source);
}