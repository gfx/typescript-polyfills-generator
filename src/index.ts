// @ts-ignore
import * as browserslist from 'browserslist';
import * as ts from 'typescript';
import normalizeBuiltInName from './normalize-built-in-name';
import semverify from './semverify';

interface BrowsersSupportList {
    // [browser name]: version
    [key: string]: string;
}

interface BuiltInsList {
    // [name]:
    [key: string]: BrowsersSupportList;
}

export type PolyfillStatusOrVersion = boolean|string|'all';

export interface PolyfillsList {
    [key: string]: PolyfillStatusOrVersion;
}

export interface PolyfillsGeneratorOptions {
    targets?: string;
    polyfills?: PolyfillsList;
}

const {hasOwnProperty} = Object.prototype;
const allBuiltInsWithEsVersion: BuiltInsList = require('../data/built-ins.json');
const allVersionsKey: string = 'all';
const allBuiltIns: BuiltInsList = Object.create(null);
const browserNameMap: {[key: string]: string} = {
    and_chr: 'chrome',
    and_ff: 'firefox',
    ios_saf: 'ios'
};

for (const builtIn in allBuiltInsWithEsVersion) {
    if (hasOwnProperty.call(allBuiltInsWithEsVersion, builtIn)) {
        const browsersList: BrowsersSupportList = allBuiltInsWithEsVersion[builtIn];
        const normalizedBrowsersList: BrowsersSupportList = Object.create(null);

        for (const browserName in browsersList) {
            if (hasOwnProperty.call(browsersList, browserName)) {
                const browserVersion: string = browsersList[browserName];

                normalizedBrowsersList[browserName] = semverify(browserVersion);
            }
        }

        allBuiltIns[normalizeBuiltInName(builtIn)] = normalizedBrowsersList;
    }
}

// export class PolyfillsGenerator {
//     private targets: string[];
//
//     constructor(options: PolyfillsGeneratorOptions) {
//         this.targets = browserslist(options.targets);
//     }
//
//     process(fileName: string, sourceText: string): Promise<string> {
//         const program: ts.Program = ts.createProgram([fileName], {});
//         const checker: ts.TypeChecker = program.getTypeChecker();
//
//         for (const sourceFile of program.getSourceFiles()) {
//             if (!sourceFile.isDeclarationFile) {
//                 ts.forEachChild(sourceFile, onNode);
//             }
//         }
//
//         function onNode(node: ts.Node) {
//             const symbol: ts.Symbol | void = checker.getSymbolAtLocation(node);
//
//             if (symbol) {
//                 console.log(
// checker.typeToString(checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration
// !)));
//             }
//
//             ts.forEachChild(node, onNode);
//         }
//
//         return Promise.resolve(sourceText);
//     }
// }

export function createPolyfillsTransformerFactory (
    program: ts.Program,
    options: PolyfillsGeneratorOptions = {}
): ts.TransformerFactory<ts.SourceFile> {
    const targets: {[key: string]: any} = Object.create(null);
    const polyfillsConfig: PolyfillsList = options.polyfills as PolyfillsList;
    const polyfills: PolyfillsList = Object.create(null);

    for (const builtIn in polyfillsConfig) {
        if (hasOwnProperty.call(polyfillsConfig, builtIn)) {
            const statusOrVersion: PolyfillStatusOrVersion = polyfillsConfig[builtIn];

            polyfills[normalizeBuiltInName(builtIn)] =
                statusOrVersion === allVersionsKey ?
                    // "first" version
                    '0.0.1' :
                    (statusOrVersion === true || statusOrVersion === false) ?
                        statusOrVersion :
                        semverify(statusOrVersion as string);

        }
    }

    browserslist(options.targets).forEach((browser: string) => {
        const [browserName, browserVersion] = browser.split(' ');
        const normalizedBrowserName: string = browserNameMap[browserName] || browserName;
        const splitVersion: string = browserVersion.split('-')[0].toLowerCase();

        targets[normalizedBrowserName] = splitVersion === allVersionsKey ? splitVersion : semverify(splitVersion);
    });

    return function polyfillsTransformerFactory (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> {
        function visitor (node: ts.Node): ts.Node {
            return ts.visitEachChild(node, visitor, context);
        }

        return function polyfillsTransformer (sourceFile: ts.SourceFile) {
            ts.visitNode(sourceFile, visitor);

            return sourceFile;
        };
    };
}