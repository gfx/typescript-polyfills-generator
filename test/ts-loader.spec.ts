import * as fs from 'fs';
import * as path from 'path';
import * as ts from 'typescript';
import {promisify} from 'util';
import {Stats} from 'webpack';
import {createPolyfillsTransformerFactory} from '../src/index';
import compile from './helpers/compiler';

const readFile = promisify(fs.readFile);

xtest(`transformer with ts-loader`, () => {
    const testFilePath: string = path.resolve(__dirname, './files/test-file-1.ts');

    return Promise.all([
        readFile(testFilePath, 'utf8'),
        compile(testFilePath, {
            tsLoaders: [
                {
                    loader: 'ts-loader',
                    options: {
                        getCustomTransformers(program: ts.Program) {
                            return {
                                before: [
                                    createPolyfillsTransformerFactory(program, {
                                        targets: 'last 2 version, not ie < 11, not ie_mob < 11, safari >= 9'
                                    })
                                ]
                            };
                        }
                    }
                }
            ]
        })
    ]).then(([source, stats]: [string, Stats]) => {
        const output: string = stats
            .toJson()
            .modules[0].source.replace(/(module\.exports = ")/, '$1\n')
            .replace(/\\n/g, '\n');

        expect(output).toBe(`module.exports = "\n${source}"`);
        expect(output).toMatchSnapshot();
    });
});
