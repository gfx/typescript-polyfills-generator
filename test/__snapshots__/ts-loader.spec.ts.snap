// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer with ts-loader 1`] = `
"module.exports = \\"'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
require('core-js/modules/es6.array.find-index.js');
require('core-js/modules/es6.array.find.js');
require('core-js/modules/es6.array.from.js');
require('core-js/modules/es6.array.iterator.js');
require('core-js/modules/es6.number.constructor.js');
require('core-js/modules/es6.number.is-nan.js');
require('core-js/modules/es6.object.assign.js');
require('core-js/modules/es6.promise.js');
require('core-js/modules/es6.set.js');
require('core-js/modules/es6.string.includes.js');
require('core-js/modules/es6.string.iterator.js');
require('core-js/modules/es7.array.includes.js');
require('core-js/modules/es7.string.pad-end.js');
require('core-js/modules/es7.string.pad-start.js');
require('typescript-polyfills-generator/lib/polyfills/es5.array.unshift.js');
require('typescript-polyfills-generator/lib/polyfills/es5.location.origin.js');
require('whatwg-fetch');
exports.elementInArray = [1, '2', 3, null].find((value) => value === 3);
exports.elementIndexInArray = [1, '2', 3, null].findIndex((value) => value === '2');
exports.hasTwoInArray = [1, '2', 3, null].includes('2');
exports.hasTwoInString = '123'.includes('2');
exports.paddedStartNumber = '9'.padStart(2, '0');
exports.paddedEndNumber = '2'.padEnd(3, '0');
exports.arrayFromString = Array.from('foo');
exports.fulfilledPromise = Promise.resolve({});
const { isNaN } = Number;
const { invalidMethod } = Array;
exports.isNotGlobalNaN = isNaN;
Object.assign({}, { value: 1 });
exports.request = fetch;
exports.locationOrigin = location.origin;
const testArr = [0, 2147483648];
testArr.shift();
testArr[1] = 1;
exports.testSet = new Set();
\\""
`;

exports[`transformer with ts-loader with no polyfills in the output 1`] = `
"module.exports = \\"'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.elementInArray = [1, '2', 3, null].find((value) => value === 3);
exports.elementIndexInArray = [1, '2', 3, null].findIndex((value) => value === '2');
exports.hasTwoInArray = [1, '2', 3, null].includes('2');
exports.hasTwoInString = '123'.includes('2');
exports.paddedStartNumber = '9'.padStart(2, '0');
exports.paddedEndNumber = '2'.padEnd(3, '0');
exports.arrayFromString = Array.from('foo');
exports.fulfilledPromise = Promise.resolve({});
const { isNaN } = Number;
const { invalidMethod } = Array;
exports.isNotGlobalNaN = isNaN;
Object.assign({}, { value: 1 });
exports.request = fetch;
exports.locationOrigin = location.origin;
const testArr = [0, 2147483648];
testArr.shift();
testArr[1] = 1;
exports.testSet = new Set();
\\""
`;
